### ISSUE TITLE
Grpc client hung during rpc request

### ISSUE THREAD

**Issue by IkshuDM** [grpc/grpc#28295]
What version of gRPC and what language are you using?
language : C++, version: 1.30.0

What operating system (Linux, Windows,...) and version?
Linux 4.19.81

What runtime / compiler are you using (e.g. python version or version of gcc)
Python 2.7.16, g++ (Debian 8.3.0-6) 8.3.0

What did you do?
When the client starts up,  2 threads are  started to  monitor the channel connectivity like below. ChannelReady thread is supposed to exit after 30 seconds  but it is hung. Any following  rpc  request will be hung. 

`
testClient:: start(const std::string& uri)
{
    gpr_log_verbosity_init();
    gpr_set_log_verbosity(GPR_LOG_SEVERITY_DEBUG);
    IsExitTh = false;
    channel = grpc::CreateChannel(uri, grpc::InsecureChannelCredentials());
    ChannelStateNotify = std::thread(&testClient::ChannelStateNotify Thread, this);

  //This is a workaround to reconnect to server
    std::thread(&testClient::ChannelReady, this, 30, true).detach();
}

void testClient::ChannelStateNotify ()
{
    bool ok = false;
    void *tag = NULL;

    while (!IsExitTh)
    {   
        auto grpc_state = channel->GetState(true);
        if(channel.get()->WaitForStateChange(
                    grpc_state, std::chrono::system_clock::now() + std::chrono::seconds(10)) == true)
        {
            std::cout <<" WaitForStateChange state changed";

        }
        else
        {
              std::cout << "WaitForStateChange timedout";
        }
    }

    std::cout << "Exit channel notification thread";
}

void testClient::ChannelReady(int timeout_seconds, bool checkStateChange) {
    grpc_connectivity_state state  = channel->GetState(false);
    while ((state != GRPC_CHANNEL_READY) && (timeout_seconds!= 0)) {
        state = channel->GetState(true /* try_to_connect */);
        if(checkStateChange)
        {
           (void)channel->WaitForStateChange(state, std::chrono::system_clock::now() + std::chrono::seconds(1));
        }
        else
        {
          sleep(1);
          std::cout << "Channel is not Ready yet";
        }
        timeout_seconds--;
    }
}
`

Please find below backtrace

Thread 59 (Thread 0x7f9ba926d700 (LWP 9468)):
#0  0x00007f9bc04f17ef in epoll_wait () from /lib/x86_64-linux-gnu/libc.so.6
#1  0x00007f9bc0fb2d67 in pollable_epoll(pollable*, long) () from libgrpc.so
#2  0x00007f9bc0fb3859 in pollset_work () from libgrpc.so
#3  0x00007f9bc0fbb0be in pollset_work () from libgrpc.so
#4  0x00007f9bc0fc25a0 in grpc_pollset_work(grpc_pollset*, grpc_pollset_worker**, long) () from libgrpc.so
#5  0x00007f9bc101ec90 in cq_next(grpc_completion_queue*, gpr_timespec, void*) () from libgrpc.so
#6  0x00007f9bc101f155 in grpc_completion_queue_next () from libgrpc.so
#7  0x00007f9bc24c0676 in grpc_impl::CompletionQueue::AsyncNextInternal(void**, bool*, gpr_timespec) () from libgrpc++.so
#8  0x00007f9bc24aeff7 in grpc_impl::CompletionQueue::Next(void**, bool*) () from libgrpc++.so
#9  0x00007f9bc24ae3b5 in grpc_impl::Channel::WaitForStateChangeImpl(grpc_connectivity_state, gpr_timespec) () from libgrpc++.so.1
#10 0x00007f9bc133538a in bool grpc::ChannelInterface::WaitForStateChange<std::chrono::time_point<std::chrono::_V2::system_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > > >(grpc_connectivity_state, std::chrono::time_point<std::chrono::_V2::system_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >) ()
#11 0x00007f9bc1333410 in testClient::ChannelReady(int, bool) () 

What did you expect to see?
ChannelReady thread should have exited after 30 second 

What did you see instead?
ChannelReady  thread if hug as  WaitForStateChange was hung at epoll_wait (). Any following rpc request done even as part of different thread will be hung like shown in the below back trace:

Thread 63 (Thread 0x7f9ba79e9700 (LWP 9737)):
#0  0x00007f9bc04f17ef in epoll_wait () from /lib/x86_64-linux-gnu/libc.so.6
#1  0x00007f9bc0fb2d67 in pollable_epoll(pollable*, long) () fromlibgrpc.so
#2  0x00007f9bc0fb3859 in pollset_work () from libgrpc.so
#3  0x00007f9bc0fbb0be in pollset_work () from libgrpc.so
#4  0x00007f9bc0fc25a0 in grpc_pollset_work(grpc_pollset*, grpc_pollset_worker**, long) () from libgrpc.so
#5  0x00007f9bc101f879 in cq_pluck(grpc_completion_queue*, void*, gpr_timespec, void*) () from libgrpc.so
#6  0x00007f9bc101faff in grpc_completion_queue_pluck () from libgrpc.so
#7  0x00007f9bc24c0a53 in grpc::CoreCodegen::grpc_completion_queue_pluck(grpc_completion_queue*, void*, gpr_timespec, void*) () from libgrpc++.so
#8  0x00007f9bc2943214 in grpc_impl::CompletionQueue::Pluck(grpc::internal::CompletionQueueTag*) () from libbase_services.so
#9  0x00007f9bc179cd34 in grpc::internal::BlockingUnaryCallImpl<Update , Update >::BlockingUnaryCallImpl(grpc::ChannelInterface*, grpc::internal::RpcMethod const&, grpc_impl::ClientContext*, Update t const&, Update *) () from 
#10 0x00007f9bc179b11d in grpc::Status grpc::internal::BlockingUnaryCall<Update ,Update>(grpc::ChannelInterface*, grpc::internal::RpcMethod const&, grpc_impl::ClientContext*, Update const&, Update*) ()

**Comment by IkshuDM** [comment]
can I please  get help with how to solve above issue

**Comment by yashykt** [comment]
Can you share your code as a gist maybe?

**Comment by IkshuDM** [comment]
> Can you share your code as a gist maybe?

Thanks for your response.  I have added the code in the issue log.

**Comment by yashykt** [comment]
Please take a look at https://github.com/grpc/grpc/blob/62c1b4f0139cd28c8c17c2f6ffefc8d391e46354/test/cpp/end2end/end2end_test.cc#L1363

The test is pretty similar to the code that you've provided and works.

From your description of the code hanging at polling and also RPCs not finishing after the deadline seems to indicate that there is something wrong with the timing system. (I'm assuming that there's nothing wrong with the rest of the code.)

To debug this further, I would suggest enabling the polling tracer - 
```
GRPC_TRACE=polling
GRPC_VERBOSITY=DEBUG
```
Note that polling is a debug mode tracer so your build would also need to have the debug option set.
The log at this line https://github.com/grpc/grpc/blob/62c1b4f0139cd28c8c17c2f6ffefc8d391e46354/src/core/lib/iomgr/ev_epollex_linux.cc#L1116 would help determine what the actual current time and deadline values are.

**Comment by yashykt** [comment]
This bug is outdated now given that the epollex polling engine has now been deleted, but if it still persists on master, please feel free to create a new issue.
